### ISSUE TITLE
Recover JMS Serializer handlers

### ISSUE THREAD

**Issue by ushatel** [FriendsOfSymfony/FOSRestBundle#1851]
Hello, Friends Of Symfony, 

Please, check this thing when this is possible.  

I am using JMS Serializer in my project with the extension jms_serializer.subscribing_handler registered 
 that allows me to make specific transformation for the type. 

After making composer updates of the vendors it won't work. 

Making some analyzes have found that the mistake rised from the \vendor\friendsofsymfony\rest-bundle\FOSRestBundle.php 

There were changes in that file  

>>> Was 

`
$container->addCompilerPass(new JMSHandlersPass(), PassConfig::TYPE_REMOVE); 
`

>>> 

>>> After upgrade 

`
$container->addCompilerPass(new JMSHandlersPass(), PassConfig::TYPE_BEFORE_REMOVING, -10); 
`

>>> 

It looks the sources of the changes are here 
https://github.com/FriendsOfSymfony/FOSRestBundle/commit/0f85a17149028f38708aca926e7eb8bbd639ca1d 
https://github.com/FriendsOfSymfony/FOSRestBundle/commit/f144dbbea507c2bca3753b96f8cf28ed82ddf550 

If I revert changes to the \vendor\friendsofsymfony\rest-bundle\FOSRestBundle.php all things goes well. 

Please, help to solve this problem. 

Thank you for the attention and wish you good time!

**Comment by xabbuh** [comment]
Can you explain what exactly does not work?

**Comment by ushatel** [comment]
The handler that installed with the  jms_serializer.subscribing_handler could not handle the type after downloading the latest FOS. When I revert changes to the previous version handle works well on the same code. 

Thank you for the attention.

**Comment by xabbuh** [comment]
Sorry, but with this little information I am afraid I will not be able to help you. Maybe it's easier to explain your problem by pushing a small example project that shows what is not working for you.

**Comment by ushatel** [comment]
Thank you for the answer. I will prepare the project as I could rise the time quote.

**Comment by naitsirch** [comment]
Same problem here. It caused me a lot of headache today. After an update of FosRestBundle, the compiled service container builds the `jms_serializer.handler_registry` this way:

```php
protected function getJmsSerializer_HandlerRegistryService()
{
    return $this->services['jms_serializer.handler_registry'] = new \FOS\RestBundle\Serializer\JMSHandlerRegistry(
        new \JMS\Serializer\Handler\LazyHandlerRegistry($this),
        array(2 => array(/* some handlers */), 1 => array(/* some handlers */))
    );
}
```
The array with the handlers isn't passed to the `LazyHandlerRegistry` anymore, but to your `JMSHandlerRegistry`, so that the handler registry does not get any handlers passed.

**Comment by naitsirch** [comment]
I have workaround this bug, by registering *FOSRestBundle* after *JMSSerializerBundle* in `AppKernel`

```php
// app/AppKernel.php
    public function registerBundles()
    {
        $bundles = array(
            // ...
            new JMS\SerializerBundle\JMSSerializerBundle(),
            new FOS\RestBundle\FOSRestBundle(), // <-- look here!!
```
@xabbuh If possible, throw an exception somewhere to inform users about this. I found this comment in [`JMSHandlersPass::registerHandlerRegistry`](https://github.com/FriendsOfSymfony/FOSRestBundle/blob/2.3.0/DependencyInjection/Compiler/JMSHandlersPass.php#L45):

> Inlined because {@link JMS\SerializerBundle\DependencyInjection\Compiler\CustomHandlersPass} must be executed before replacing jms_serializer.handler_registry.

This made me try to place the serializer bundle before the rest bundle.

**Comment by Tobion** [comment]
@xabbuh without https://github.com/FriendsOfSymfony/FOSRestBundle/commit/f144dbbea507c2bca3753b96f8cf28ed82ddf550 (which is not tagged unfortuantely) the rest bundle with SF 4 has broken exception serialization. It cannot find the `jms_serializer.handler_registry` so it removes the `fos_rest.serializer.exception_normalizer.jms`. Without that normalizer, the serialization of the errors is different and it even serializes the backtrace. This has be considered a security hole.
So it would be important to resolve this issue and make a release.

**Comment by xabbuh** [comment]
@Tobion Sadly, exactly that commit broke our test suite (I didn't notice before merging), see https://travis-ci.org/FriendsOfSymfony/FOSRestBundle/jobs/317959484. So I am a bit reluctant in making a tag in this state.

**Comment by xabbuh** [comment]
should be fixed by #1873

**Comment by xabbuh** [comment]
To be more clear: updating to 2.3.1 should help you
